op-e2e/actions/l1_miner.go

```diff
@@ -0,0 +1,142 @@
+package actions
+
+import (
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/consensus/misc"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/trie"
+)
+
+// L1Miner wraps a L1Replica with instrumented block building ability.
+type L1Miner struct {
+	L1Replica
+
+	// L1 block building data
+	l1BuildingHeader *types.Header             // block header that we add txs to for block building
+	l1BuildingState  *state.StateDB            // state used for block building
+	l1GasPool        *core.GasPool             // track gas used of ongoing building
+	pendingIndices   map[common.Address]uint64 // per account, how many txs from the pool were already included in the block, since the pool is lagging behind block mining.
+	l1Transactions   []*types.Transaction      // collects txs that were successfully included into current block build
+	l1Receipts       []*types.Receipt          // collect receipts of ongoing building
+	l1Building       bool
+	l1TxFailed       []*types.Transaction // log of failed transactions which could not be included
+}
+
+// NewL1Miner creates a new L1Replica that can also build blocks.
+func NewL1Miner(log log.Logger, genesis *core.Genesis) *L1Miner {
```

Comment:
Not super opinionated but should this be "block producer"?

Reply:
The Miner name was used in the code before the Merge, haha :sweat_smile: . It's also making them canonical, so maybe it's just L1Sequencer ? I still like the distinction with the L2 naming.

Reply:
Think I'll leave it as is-is for now and avoid rebasing other in-flight work for the rename. But happy to rename later if we have a better name for this actor.

Reply:
Also, this is part of the other action testing which was merged in develop, but the mergify squashing caused this stacked PR to contain old commits.